### DISCUSSION TITLE
Best Practices for Debugging IK Failures

### DISCUSSION THREAD

**Discussion by branvu** [NVlabs/curobo#158] [Software Q&A] [answered]
Hello!

Do you have recommendations for debugging why IK might be failing. I'm currently using the python library with Isaac gym. I convert my object poses (box on table) to the worldConfig and then I'm running 

`ik_config = IKSolverConfig.load_from_robot_config(
            robot_cfg,
            self.make_world(),
            rotation_threshold=0.01,
            position_threshold=0.03,
            num_seeds=500,
            self_collision_check=False,
            self_collision_opt=False,
            tensor_args=TensorDeviceType("cuda:0"),
            use_cuda_graph=True,
            # use_fixed_samples=True,
        )`

I'm then setting the joint state directly in isaac gym to the result from the IK. In most of my cases, I'm seeing that the orientation is very off. I even turned off all collisions other than the gripper as well. I don't think my target is in collision because i'm marking it on isaac gym and it's not in collision. I'm also transforming my point+all objects from isaac gym's world frame to my robot base frame. I know my target location and orientation is reachable because I can use impedence control to reach it. I'm wondering if there is any log messages or things I should be looking for in order to debug this further. 

There is one orientation + point that does work and it happens to be where the gripper is only rotated 90 along one axis which may make the convergence easier? Though i'd expect it still would be able to converge for the other cases too.

**Comment by branvu** [comment]
Ok, update on my specific problem. It turned out quaternions in curobo are w,x,y,z and isaac gym likes them as x,y,z,w so that was my source of error. Though I do have an additional question:

- Does commenting out the collision tag in the urdf actually turn off collisions for those links during planning?
- Also, I want the target to be the tip of the gripper not the hand, is there a recommended way to set this?

**Reply by balakumar-s** [reply, marked as answer]
1. collision tag is not read by cuRobo. 
2. you need to add the tip as an additional link in the urdf and then set ee_link to this frame in the robot configuration file.

**Reply by branvu** [reply]
Is there any way to turn off collision of certain links?

**Reply by balakumar-s** [reply]
Yes, you can remove the link from the list here: https://github.com/NVlabs/curobo/blob/88eac64edca2d2b641c1622e63f3e750157596c4/src/curobo/content/configs/robot/franka.yml#L37
